### ISSUE TITLE
Fix FP S2259 (`null-dereference`)

### ISSUE THREAD

**Issue by vilchik-elena** [SonarSource/SonarJS#3124]
Check issues on Peach to see those FPs

**Comment by victor-diez-sonarsource** [comment]
After some research, seems the typechecker reports undefined when values are initialized to this.attr. Example to reproduce:

```js
function test() {
  var foo = this.bar;
  foo = this.bar = 1;
  console.log(foo.length);
}
```
However, all these will not be reported:

```js
function test() {
  var foo = this.bar;
  foo = this.bar;
  console.log(foo.length);
}
```
```js
function test() {
  var foo;
  foo = this.bar = 1;
  console.log(foo.length);
}
```
```js
function test() {
  var foo = bar;
  foo = bar = 1;
  console.log(foo.length);
}
```

**Comment by ilia-kebets-sonarsource** [comment]
Very weird, this triggers:
```javascript
function test() {
  let foo = this.bar
  foo = this.bar = 1;
  console.log('salut', foo.length);
}
```
This doesn't (removed redundant line 2):
```javascript
function test() {
  let foo = this.bar = 1;
  console.log('salut', foo.length);
}
```

**Comment by ilia-kebets-sonarsource** [comment]
Here are the cases I identified on peach, but the simplified versions do not work: https://gist.github.com/ilia-kebets-sonarsource/9baf23abb45eb095f5402bb0a657aec5

**Comment by ilia-kebets-sonarsource** [comment]
It seems that this issue might have been partially fixed in the meantime, since running an analysis with SQ 9.5 on the same file as this issue raises nothing: https://peach.sonarsource.com/issues?resolved=false&rules=javascript%3AS2259&open=AYPEy2PDkcjc1EUKVg5U

**Comment by ilia-kebets-sonarsource** [comment]
With the current rule implementation, it's hard to change the behaviour of the rule, as we rely on the TypeScript TypeChecker. The number of issues the rule produces is low (~50 issues on Peach on the moment of writing), so even if the FP rate is high, the impact is negligible.
So, as of today, we decided not to fix these FPs.

**Comment by ilia-kebets-sonarsource** [comment]
closing as won't fix
